### PR TITLE
Clean up `network`; refine `encode` types, update deps

### PR DESCRIPTION
**Pull request by cryptodeal**
General cleanup/update dependencies. This fixes the type errors in a few of the examples due to `grad` related API changes.

### CODE REVIEW

File: shumai/tensor/tensor.ts [hunk 1]
```diff
@@ -416,7 +416,7 @@ export class Tensor {
     }
   }
 
-  checkpoint(file?: () => any, callback?: () => boolean) {
+  checkpoint(file?: () => any | any, callback?: () => boolean) {
```
Comment (asilvas) [inline shumai/tensor/tensor.ts:419]: mistake or is this doing some voodoo I'm not aware of?

Comment (cryptodeal) [reply shumai/tensor/tensor.ts:419]: Ahh, good catch! Had been toying with trying to name all possible types, but yea that's an error. I'll fix that now!

File: examples/distributed/model_a.ts [hunk 2]
```diff
@@ -21,4 +21,4 @@ class MLP extends sm.module.Module {
 
 const model = new MLP()
 
-sm.network.serve_model(model, (ts) => sm.optim.sgd(ts, 1e-4), { port: 3001 })
+sm.network.serve_model(model.forward, (ts) => sm.optim.sgd(ts, 1e-4), { port: 3001 })
```
Comment (bwasti) [inline examples/distributed/model_a.ts:24]: this should be fine, is it a type error?

Comment (cryptodeal) [reply examples/distributed/model_a.ts:24]: Yea, it was a type error; `serve_model` expects the first arg to be `(t: sm.Tensor) => sm.Tensor | Promise<sm.Tensor>`

File: shumai/tensor/tensor.ts (hunk 1 above)
Comment (cryptodeal) [reply shumai/tensor/tensor.ts]: Fixed; changed the conditional to `if (file instanceof Function)` as that should be marginally more performant.

File: examples/distributed/model_a.ts (hunk 2 above)
Comment (cryptodeal) [reply examples/distributed/model_a.ts:24]: I can update the argument type (and make any needed changes in the logic) in `serve_model` though if that's the route to go here

**Review by asilvas — APPROVED** [review]

**Review by bwasti — APPROVED** [review]

**Pull request merged**